### PR TITLE
[Enhancement] support schema scan in pipeline engine

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/SchemaScanNode.java|0|1|00.00%|[144]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalSchemaScanOperator.java|1|1|100.00%|[]|

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/SchemaScanNode.java|0|1|00.00%|[144]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalSchemaScanOperator.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

File: be/src/exec/vectorized/schema_scan_node.cpp [hunk 1]
```diff
@@ -283,4 +288,19 @@ Status SchemaScanNode::set_scan_ranges(const std::vector<TScanRangeParams>& scan
     return Status::OK();
 }
 
+std::vector<std::shared_ptr<pipeline::OperatorFactory>> SchemaScanNode::decompose_to_pipeline(
+        pipeline::PipelineBuilderContext* context) {
+    auto* morsel_queue_factory = context->morsel_queue_factory_of_source_operator(id());
+    size_t dop = morsel_queue_factory->size();
+
+    size_t buffer_capacity = pipeline::ScanOperator::max_buffer_capacity() * dop;
```
Comment (kangkaisen) [inline be/src/exec/vectorized/schema_scan_node.cpp:296]: in non-pipeline engine, the degree of parallelism of schema scan should be 1.

File: be/src/exec/pipeline/scan/olap_schema_chunk_source.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    _scanner = vectorized::SchemaScanner::create(schema_table->schema_table_type());
+    if (_scanner == nullptr) {
+        return Status::InternalError("schema scanner get null pointer");
+    }
+
+    RETURN_IF_ERROR(_scanner->init(param, _ctx->object_pool()));
+    return _scanner->start(state);
+}
+
+void OlapSchemaChunkSource::close(RuntimeState* state) {}
```
Comment (kangkaisen) [inline be/src/exec/pipeline/scan/olap_schema_chunk_source.cpp:46]: don't need we need to do anything when close?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/SchemaScanNode.java|0|1|00.00%|[144]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalSchemaScanOperator.java|1|1|100.00%|[]|

File: be/src/exec/pipeline/scan/olap_schema_chunk_source.cpp (hunk 2 above)
Comment (silverbullet233) [reply be/src/exec/pipeline/scan/olap_schema_chunk_source.cpp:46]: > don't need we need to do anything when close?

Yes, we don't need to do anything actively, using the destructor to release resources is enough.

File: be/src/exec/vectorized/schema_scan_node.cpp (hunk 1 above)
Comment (silverbullet233) [reply be/src/exec/vectorized/schema_scan_node.cpp:296]: > in non-pipeline engine, the degree of parallelism of schema scan should be 1.

fixed…In fact, `morsel_queue_factory->size()` will only return 1 for SchemaScanOperator, but I think it's more appropriate to explicitly set dop to 1.

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

File: be/src/exec/vectorized/schema_scan_node.cpp [hunk 3]
```diff
@@ -283,4 +288,19 @@ Status SchemaScanNode::set_scan_ranges(const std::vector<TScanRangeParams>& scan
...
 
+std::vector<std::shared_ptr<pipeline::OperatorFactory>> SchemaScanNode::decompose_to_pipeline(
+        pipeline::PipelineBuilderContext* context) {
+    // the dop of SchemaScanOperator should always be 1.
+    size_t dop = 1;
+
+    size_t buffer_capacity = pipeline::ScanOperator::max_buffer_capacity() * dop;
+    int64_t mem_limit = runtime_state()->query_mem_tracker_ptr()->limit() * config::scan_use_query_mem_ratio;
+    pipeline::ChunkBufferLimiterPtr buffer_limiter = std::make_unique<pipeline::DynamicChunkBufferLimiter>(
+            buffer_capacity, buffer_capacity, mem_limit, runtime_state()->chunk_size());
```
Comment (fzhedu) [inline be/src/exec/vectorized/schema_scan_node.cpp:299]: is there any check of validation for `config::scan_use_query_mem_ratio`?

File: be/src/exec/vectorized/schema_scan_node.cpp [hunk 4]
```diff
@@ -14,11 +16,14 @@ namespace starrocks::vectorized {
...
         : ScanNode(pool, tnode, descs),
+          _tnode(tnode),
           _is_init(false),
           _table_name(tnode.schema_scan_node.table_name),
           _tuple_id(tnode.schema_scan_node.tuple_id),
           _dest_tuple_desc(nullptr),
-          _schema_scanner(nullptr) {}
+          _schema_scanner(nullptr) {
+    _name = "schema_scan";
+}
```
Comment (fzhedu) [inline be/src/exec/vectorized/schema_scan_node.cpp:26]: _schema_scanner(nullptr),
_name(schema_scan) {
}

File: be/src/exec/pipeline/scan/olap_schema_chunk_source.cpp (hunk 2 above)
Comment (fzhedu) [reply be/src/exec/pipeline/scan/olap_schema_chunk_source.cpp:46]: when the instance is reused, but we should reset its status, so close() shoulc clear some context.

File: be/src/exec/pipeline/scan/olap_schema_scan_context.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    if (_tnode.schema_scan_node.__isset.ip) {
+        _param->ip = _obj_pool.add(new std::string(_tnode.schema_scan_node.ip));
+    }
+    if (_tnode.schema_scan_node.__isset.port) {
+        _param->port = _tnode.schema_scan_node.port;
+    }
+
+    if (_tnode.schema_scan_node.__isset.thread_id) {
+        _param->thread_id = _tnode.schema_scan_node.thread_id;
+    }
```
Comment (fzhedu) [inline be/src/exec/pipeline/scan/olap_schema_scan_context.cpp:54]: should we valid the parameters?

File: be/src/exec/pipeline/scan/olap_schema_scan_context.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    ~OlapSchemaScanContext() = default;
+
+    Status prepare(RuntimeState* state);
+
+    BalancedChunkBuffer& get_chunk_buffer() { return _chunk_buffer; }
+
+    TupleId tuple_id() { return _tuple_id; }
+
+    ObjectPool* object_pool() { return &_obj_pool; }
```
Comment (fzhedu) [inline be/src/exec/pipeline/scan/olap_schema_scan_context.h:39]: * const object_pool()

File: be/src/exec/pipeline/scan/olap_schema_chunk_source.cpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        {
+            SCOPED_TIMER(_filter_timer);
+            auto& conjunct_ctxs = _ctx->conjunct_ctxs();
+            if (!conjunct_ctxs.empty()) {
+                RETURN_IF_ERROR(ExecNode::eval_conjuncts(conjunct_ctxs, chunk_dst.get()));
+            }
+        }
+        row_num = chunk_dst->num_rows();
```
Comment (fzhedu) [inline be/src/exec/pipeline/scan/olap_schema_chunk_source.cpp:86]: what if `row_num == 0 ?`

File: be/src/exec/pipeline/scan/olap_schema_chunk_source.cpp (hunk 2 above)
Comment (silverbullet233) [reply be/src/exec/pipeline/scan/olap_schema_chunk_source.cpp:46]: > when the instance is reused, but we should reset its status, so close() shoulc clear some context.

It seems that ChunkSource won't be reused, in scan operator, we will create a new ChunkSource for each morsel.

File: be/src/exec/vectorized/schema_scan_node.cpp (hunk 4 above)
Comment (silverbullet233) [reply be/src/exec/vectorized/schema_scan_node.cpp:26]: > _schema_scanner(nullptr), _name(schema_scan) { }

_name is a member of the parent class ScanNode. It seems that we can't put it in the initialization list.

File: be/src/exec/vectorized/schema_scan_node.cpp (hunk 3 above)
Comment (silverbullet233) [reply be/src/exec/vectorized/schema_scan_node.cpp:299]: > is there any check of validation for `config::scan_use_query_mem_ratio`?

I'm afraid there isn't. I think the verification of configurations should be handled centrally when the BE process starts. we can improve it later

File: be/src/exec/pipeline/scan/olap_schema_scan_context.cpp (hunk 5 above)
Comment (silverbullet233) [reply be/src/exec/pipeline/scan/olap_schema_scan_context.cpp:54]: > should we valid the parameters?

These will be verfied in `SchemaScanner::start`, I think we needn't valid them here.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/SchemaScanNode.java|0|1|00.00%|[144]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalSchemaScanOperator.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Review by kangkaisen — DISMISSED** [review]

**Review by fzhedu — DISMISSED** [review]
LGTM

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by murphyatwork — DISMISSED** [review]

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/SchemaScanNode.java|0|1|00.00%|[144]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalSchemaScanOperator.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Review by murphyatwork — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**